src/Microsoft.Kiota.Cli.Commons/Http/Headers/InMemoryHeadersStore.cs

```diff
@@ -0,0 +1,55 @@
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.Kiota.Cli.Commons.Http.Headers;
+
+/// <summary>
+/// An in memory headers store.
+/// </summary>
+/// <remarks>
+/// This class is not thread safe.
+/// </remarks>
+public sealed class InMemoryHeadersStore : BaseHeadersStore
+{
+    private readonly Dictionary<string, ICollection<string>> _headers = new();
+
+    /// <summary>
+    /// Returns a single instance of the <see cref="InMemoryHeadersStore"/>
+    /// singleton.
+    /// </summary>
+    /// <remarks>
+    /// For cases where you might not want to use a single instance across your
+    /// application e.g. when using an instance per thread, create a new
+    /// instance and use that instead. In that case, make sure you have cached
+    /// the instance when you need to access the stored headers.
+    /// </remarks>
+    public static InMemoryHeadersStore Instance { get; } = new InMemoryHeadersStore();
+
+    /// <inheritdoc />
+    public override IEnumerable<KeyValuePair<string, ICollection<string>>> Headers() => _headers.AsEnumerable();
+
+    /// <inheritdoc />
+    public override IEnumerable<KeyValuePair<string, ICollection<string>>> Drain()
+    {
+        var existing = new List<KeyValuePair<string, ICollection<string>>>(_headers.Count);
+        existing.AddRange(_headers);
+        _headers.Clear();
+        return existing;
+    }
+
+    /// <inheritdoc />
+    protected override void DoAddHeader(string name, string value)

```

Comment:
why not simply add header as a name?

Reply:
What do you mean?

Reply:
DoAddHeader seems like an odd name. Why not simply call it AddHeader?

Reply:
Because it's an unchecked private member. The convention is intentional to tell people that it's part of a larger routine. I've seen this pattern in other libraries like MSAL
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/80c848b5b7bfc033b11ee82107344bfc22efc0da/src/client/Microsoft.Identity.Client/json/JsonTextWriter.Async.cs#L60

Reply:
I'd like to revisit the need for a concurrent dictionary here, why do we think one is needed? To Caleb's comment the process is short lived, runs in a single thread.

If we do, providing two API surfaces for storage of the Headers is a poor developer experience. We should at the very least change the property type to IDictionary and encapsulate the instance type.